### PR TITLE
Added menubar and buttons to the GUI

### LINKED ISSUE #6: Create dummy buttons
# Add basic buttons to the GUI

The GUI should have the following buttons:
- [x] **File** for opening and saving files
- [x] **View** for light and darkmode
- [x] **Polynomials** for choosing the number of used polynomials, as well as equidistant or exponential distance
- [x] **Alpha calculation** for choosing between the options of calculating alpha (Sum Ratio or Weighted Mean)

Unless otherwise possible, the buttons serve as dummies, functionality should be implemented later on.

Additionaly to that two rows of buttons should be implemented.
One for choosing which data points to use for the fit, the other one for choosing which data points to use for the calculation of &tau; and &Delta; &tau;.

### PR DESCRIPTION
**Pull request by julius-kress**
### Summary

This adds a menu bar to the GUI, as well as checkboxes for selecting which datapoints to use for the fitting of the graph and the calculation of tau.
It also replaces the previous UI mockup.

---

### PR checklist

- [X] I have written tests that fail without my changes. _Or I did not add them on purpose._
- [X] I have included instructions on how to test this, assuming they are not already apparent from the linked issue.

---

## How to test

There are no end-to-end tests for the UI yet, therefore you can only check the GUI manually.
You can check if it is working properly by making sure it starts without errors.
All buttons should be able to be pressed without errors.
Buttons where the backend functionality is not yet implemented will at least give an output to the developer terminal.

### Closes/Fixes/Related to

* Closes #6

### CODE REVIEW

File: napytau/gui/gui_app.py [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+def init(cli_arguments: CLIArguments) -> None:
+    app = GUIApp()
+    app.mainloop()
+
+def plot(value: int, window: GUIApp) -> Canvas:
+    """
+    Plot the graph.
+    :param value: The value.
+    :param window: The window.
+    :return: The canvas.
```
Comment (Benedikt-Brunner) [inline napytau/gui/gui_app.py:459]: As noted above this comment seems trivial

File: napytau/gui/gui_app.py [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+
+    # plotting the graph
+    plot1.plot(y)
+
+    # creating the Tkinter canvas
+    # containing the Matplotlib figure
+    canvas = FigureCanvasTkAgg(fig, master=window)
+    canvas.draw()
+
+    return canvas.get_tk_widget()
```
Comment (Benedikt-Brunner) [inline napytau/gui/gui_app.py:478]: I will leave these comments to you, they don't seem terribly trivial but we don't really need all of them, do as you see fit.

**Review by Benedikt-Brunner — CHANGES_REQUESTED** [review]

**Comment by Benedikt-Brunner** [comment]
I forgot to write it in the final review comment, but the GUI looks really nice, good job!
I think after we make the code a bit more modular this will be really great 👍

File: napytau/gui/gui_app.py [hunk 3]
```diff
@@ -0,0 +1,478 @@
+from typing import List, Tuple
+
```
Comment (Benedikt-Brunner) [inline napytau/gui/gui_app.py:2]: There aren't any Tests for this, please add some. If you aren't sure how or what to test, we can try to come up with some guidelines.

File: napytau/gui/gui_app.py [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+        self._init_graph()
+
+        # Initialize Checkboxes with datapoints
+        self._init_datapoint_checkboxes()
+
+        # Initialize control area
+        self._init_control_area()
+
+        # Initialize information area
+        self._init_information_area()
```
Comment (Benedikt-Brunner) [inline napytau/gui/gui_app.py:81]: Area is a really general term, as is information. We might consider renaming this to `control_panel` and `log_list`.

Comment (Benedikt-Brunner) [inline napytau/gui/gui_app.py:81]: There is an open draft that i thought was for adding this, you might want to split this out into a separate PR and attach that to the other Issue.

**Review by Benedikt-Brunner — CHANGES_REQUESTED** [review]
Noticed some more things, primarily that you did more work than the Issue describes and we might want to attach that to the other Issue for tracking (very cool, that it's already done however 😅)

File: napytau/gui/checkbox_datapoint.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
+class CheckboxDataPoint:
+    def __init__(self, coordinates: tuple[float, float], is_checked: bool):
+        """
+        Constructor for the CheckboxDataPoint class.
+
+        :param coordinates: Tuple of the coordinates.
+        :param is_checked: Boolean value if the checkbox is checked.
+        """
```
[the remark this answers is not in the document]
Comment (julius-kress) [reply napytau/gui/checkbox_datapoint.py]: I rephrased the comment.

File: napytau/gui/checkbox_datapoint.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
+class CheckboxDataPoint:
```
[the remark this answers is not in the document]
Comment (julius-kress) [reply napytau/gui/checkbox_datapoint.py]: Done.

File: napytau/gui/checkbox_datapoint.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+        :param coordinates: Tuple of the coordinates.
+        :param is_checked: Boolean value if the checkbox is checked.
+        """
+        if not isinstance(coordinates, tuple) or len(coordinates) != 2:
+            raise ValueError("coordinates must be a tuple of two float values")
+        if not all(isinstance(coord, float) for coord in coordinates):
+            raise TypeError("all coordinates must be floats.")
+        if not isinstance(is_checked, bool):
+            raise TypeError("is_checked must be a bool.")
```
[the remark this answers is not in the document]
Comment (julius-kress) [reply napytau/gui/checkbox_datapoint.py]: You're right that validation does not have to happen here.

File: napytau/gui/checkbox_datapoint.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+        self.coordinates = coordinates
+        self.is_checked = is_checked
+
+    def toggle_checkbox(self) -> None:
+        """
+        Toggles the boolean value
+        :return: None
+        """
```
[the remark this answers is not in the document]
Comment (julius-kress) [reply napytau/gui/checkbox_datapoint.py]: I rephrased the comment.

File: napytau/gui/checkbox_datapoint.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+            raise ValueError("coordinates must be a tuple of two float values")
+        if not all(isinstance(coord, float) for coord in coordinates):
+            raise TypeError("all coordinates must be floats.")
+        if not isinstance(is_checked, bool):
+            raise TypeError("is_checked must be a bool.")
+
+        self.coordinates = coordinates
+        self.is_checked = is_checked
+
+    def toggle_checkbox(self) -> None:
```
[the remark this answers is not in the document]
Comment (julius-kress) [reply napytau/gui/checkbox_datapoint.py]: Sure, i renamed the method.

File: napytau/gui/gui_app.py [hunk 10]
```diff
@@ -0,0 +1,478 @@
...
+
+# Modes: "System" (standard), "Dark", "Light"
+customtkinter.set_appearance_mode("System")
+# Themes: "blue" (standard), "green", "dark-blue"
+customtkinter.set_default_color_theme("blue")
+
+class GUIApp(customtkinter.CTk):
+    def __init__(self) -> None:
+        """
+        Constructor for the GUIApp, initializes the GUI.
```
[the remark this answers is not in the document]
Comment (julius-kress) [reply napytau/gui/gui_app.py:33]: I did this on purpose because I didn't want to over-modularize the code.
If for example the menubar is a separate class, one could argue to also refactor the buttons within the menubar (like File, View, ...) into separate classes and also the options within these buttons.
Then you end up with a lot of classes which are all initialized and used once in the same spot.

But I also see your point, especially for the graph.
What is your opinion on splitting up the GUI only into the classes MenuBar, Graph, Informationfield and Checkboxfield (names more or less like that)?
I would refactor the menubar and checkboxes which this PR/Issue is about and leave the graph and info-field for a separate Issue/PR.

File: napytau/gui/gui_app.py (hunk 2 above)
Comment (julius-kress) [reply napytau/gui/gui_app.py:479]: Same as above, this is some leftover code from the graph mockup and will be replaced soon.

File: napytau/gui/gui_app.py (hunk 4 above)
Comment (julius-kress) [reply napytau/gui/gui_app.py:82]: I like these suggestions and noted them. But I think this should be part of separate issues.

Comment (julius-kress) [reply napytau/gui/gui_app.py:82]: This PR only adds the buttons like described in the related issue. However I had to move/replace the previous mockup in some places.
But logically the graph, the control panel and the log list are not implemented yet.

Comment (julius-kress) [reply napytau/gui/gui_app.py:82]: I created the issues #18 , #19 , #20 for that.

File: napytau/gui/app.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+        }
+
+        # Initialize the menu bar
+        self.menu_bar = MenuBar(self, menu_bar_callbacks)
+
+        # Initialize the graph
+        self.graph = Graph(self)
+
+        # Initialize the checkbox panel
+        self.checkbox_panel = CheckboxPanel(self)
```
Comment (Benedikt-Brunner) [inline napytau/gui/app.py:80]: Simply lovely

File: napytau/gui/components/checkbox_panel.py [hunk 12]
```diff
@@ -0,0 +1,106 @@
+import customtkinter
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from napytau.gui.app import App  # Import only for the type checking.
```
Comment (Benedikt-Brunner) [inline napytau/gui/components/checkbox_panel.py:5]: Props for going above and beyond for static analysis 💪

**Review by Benedikt-Brunner — APPROVED** [review]
Muy bueno, there were a few comments where i would have adjusted the wording, but let's just make an issue for reviewing all comments across the repository and push that towards the end of the project. For now it's much more important to get this huge chunk of code into the repo and take that momentum into the next iteration. 👍 

image
